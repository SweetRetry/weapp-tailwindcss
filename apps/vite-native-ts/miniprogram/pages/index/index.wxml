<view class="p-4">
  <IceButton class="bg-[#d72929]" my-class="bg-[orange]"></IceButton>
</view>