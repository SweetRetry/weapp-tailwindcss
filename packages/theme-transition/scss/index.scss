@use './mixins';
@include mixins.theme-transition;
