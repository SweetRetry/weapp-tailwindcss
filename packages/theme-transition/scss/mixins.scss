@mixin theme-transition($darkSelector: '.dark') {
  ::view-transition-old(root),
  ::view-transition-new(root) {
    mix-blend-mode: normal;
    animation: none;
  }

  ::view-transition-old(root) {
    z-index: 1;
  }

  ::view-transition-new(root) {
    z-index: 2147483646;
  }

  #{$darkSelector}::view-transition-old(root) {
    z-index: 2147483646;
  }

  #{$darkSelector}::view-transition-new(root) {
    z-index: 1;
  }
}
