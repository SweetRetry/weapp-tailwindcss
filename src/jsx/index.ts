import { parse, traverse, generate } from '@/babel'
import type { ASTReplacer } from './replacer'
import { replaceWxml } from '@/wxml'

export function jsxHandler (rawSource: string, replacer: ASTReplacer) {
  // https://github.com/sonofmagic/weapp-tailwindcss-webpack-plugin/issues/53
  replacer?.end()
  const ast = parse(rawSource, {
    sourceType: 'unambiguous'
  })

  traverse(ast, {
    enter (path) {
      replacer.transform(path)
    },
    noScope: true
  })

  return generate(ast)
}

export function newJsxHandler (rawSource: string) {
  const ast = parse(rawSource, {
    sourceType: 'unambiguous'
  })
  // https://astexplorer.net/
  // TODO
  let startFlag = false
  traverse(ast, {
    ObjectProperty: {
      enter (path) {
        if (path.node.key.type === 'Identifier' && ['className', 'hoverClass', 'hoverClassName'].includes(path.node.key.name)) {
          startFlag = true
        }
      },
      exit (path) {
        if (path.node.key.type === 'Identifier' && ['className', 'hoverClass', 'hoverClassName'].includes(path.node.key.name)) {
          startFlag = false
        }
      }
    },
    StringLiteral: {
      enter (path, state) {
        if (startFlag) {
          path.node.value = replaceWxml(path.node.value, {
            keepEOL: true
          })
        }
      }
    },
    noScope: true
  })

  return generate(ast)
}
