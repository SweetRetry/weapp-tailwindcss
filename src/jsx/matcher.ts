import type { Node, ObjectProperty, Identifier, StringLiteral, Expression } from '@babel/types'

export type UserMatchNode = ObjectProperty & { key: Identifier | StringLiteral }

export function isSpecNode (node: Node) {
  return node.type === 'ObjectProperty' && node.key.type === 'Identifier'
}

export function getKey (node: Identifier | StringLiteral | Expression): string {
  if (node.type === 'Identifier') {
    return node.name
  }
  if (node.type === 'StringLiteral') {
    return node.value
  }
  return ''
}

export function reactMatcher (node: UserMatchNode) {
  // rax ->class and className, taro 为 hoverClass remax 为 hoverClassName
  //  node.key.name === 'class' || node.key.name === 'className' || node.key.name === 'hoverClass' || node.key.name === 'hover-class' || node.key.name === 'hoverClassName'
  // react jsx and vue jsx
  return ['className', 'hoverClass', 'hoverClassName', 'class', 'hover-class'].includes(getKey(node.key))
}

export function vue2Matcher (node: UserMatchNode): [boolean, UserMatchNode] {
  // 应该获取指定的节点开始匹配而不是整个 attrs 节点
  // for hover-class
  if (getKey(node.key) === 'attrs' && node.type === 'ObjectProperty') {
    if (node.value.type === 'ObjectExpression' && Array.isArray(node.value.properties)) {
      const idx = node.value.properties.findIndex((x) => {
        return x.type === 'ObjectProperty' && x.key.type === 'StringLiteral' && x.key.value === 'hover-class'
      })
      if (idx > -1) {
        return [true, node.value.properties[idx] as UserMatchNode]
      } else {
        return [false, node]
      }
    }
    return [false, node]
  }
  return [getKey(node.key) === 'class' || getKey(node.key) === 'staticClass', node] // || node.key.name === 'hover-class'
}

export function vue3Matcher (node: ObjectProperty & { key: Identifier | StringLiteral }) {
  if (['class', 'hover-class'].includes(getKey(node.key))) {
    return true
  }
  return false
}

export function isVue3SpecNode (node: Node) {
  return node.type === 'ObjectProperty' && (node.key.type === 'Identifier' || node.key.type === 'StringLiteral')
}
