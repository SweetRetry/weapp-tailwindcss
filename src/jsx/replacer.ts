import type { Node, ObjectProperty, Identifier } from '@babel/types'
import type { NodePath } from '@babel/traverse'
import { replaceWxml } from '@/wxml'
import { classStringReplace } from '@/reg'

export type UserMatchNode = ObjectProperty & { key: Identifier }
export type Replacer = {
  (path: NodePath<Node>): void
  start: (node: Node) => void
  end: () => void
}

function isSpecNode (node: Node) {
  return node.type === 'ObjectProperty' && node.key.type === 'Identifier'
}

function reactMatcher (node: UserMatchNode) {
  return node.key.name === 'className' || node.key.name === 'hoverClass'
}

function vue2Matcher (node: UserMatchNode) {
  return node.key.name === 'class' || node.key.name === 'staticClass' // || node.key.name === 'hover-class'
}

export type JsxFrameworkEnum = 'react' | 'vue' | 'vue2' | 'vue3'

// default react
export function createReplacer (framework: string = 'react'): Replacer {
  let classObjectNode: Node | null
  let startFlag = false
  const isVue3 = framework === 'vue3'
  const isVue2 = framework === 'vue' || framework === 'vue2'
  const isReact = framework === 'react'

  function start (node: Node) {
    startFlag = true
    classObjectNode = node
  }

  function end () {
    startFlag = false
    classObjectNode = null
  }

  if (isVue2) {
    const replacer = (path: NodePath<Node>) => {
      // vue2
      if (isSpecNode(path.node) && vue2Matcher(path.node as UserMatchNode)) {
        return start(path.node)
      }

      if (startFlag) {
        if ((path.node.start as number) > ((classObjectNode as Node).end as number)) {
          return end()
        }
        if (path.node.type === 'StringLiteral') {
          // TODO
          // 现在这样是有个问题的,变量中用户使用了 'a/s' 就会产生破坏效果
          path.node.value = replaceWxml(path.node.value, true)
        }
      }
    }
    replacer.start = start
    replacer.end = end
    return replacer
  } else if (isVue3) {
    const replacer = (path: NodePath<Node>) => {
      if (
        path.node.type === 'CallExpression' &&
        path.node.arguments.length === 2 &&
        path.node.arguments[0].type === 'StringLiteral' &&
        path.node.arguments[1].type === 'NumericLiteral'
      ) {
        path.node.arguments[0].value = classStringReplace(path.node.arguments[0].value, (x) => {
          return replaceWxml(x, true)
        })
      }

      if (isSpecNode(path.node) && vue2Matcher(path.node as UserMatchNode)) {
        return start(path.node)
      }

      if (startFlag) {
        if ((path.node.start as number) > ((classObjectNode as Node).end as number)) {
          return end()
        }
        if (path.node.type === 'StringLiteral') {
          path.node.value = replaceWxml(path.node.value, true)
        }
      }
    }
    replacer.start = start
    replacer.end = end
    return replacer
  } else if (isReact) {
    const replacer = (path: NodePath<Node>) => {
      // react
      if (isSpecNode(path.node) && reactMatcher(path.node as UserMatchNode)) {
        return start(path.node)
      }

      if (startFlag) {
        if ((path.node.start as number) > ((classObjectNode as Node).end as number)) {
          return end()
        }
        if (path.node.type === 'StringLiteral') {
          path.node.value = replaceWxml(path.node.value, true)
        }
      }
    }
    replacer.start = start
    replacer.end = end
    return replacer
  } else {
    const replacer = (path: NodePath<Node>) => {}
    replacer.start = start
    replacer.end = end
    return replacer
  }
}
