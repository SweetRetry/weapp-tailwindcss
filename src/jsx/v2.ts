import { replaceWxml } from '@/wxml'
import { parse, traverse, generate } from '@/babel'

export function newJsxHandler (rawSource: string) {
  const ast = parse(rawSource, {
    sourceType: 'unambiguous'
  })
  // https://astexplorer.net/
  // TODO
  let startFlag = false
  traverse(ast, {
    ObjectProperty: {
      enter (path) {
        if (path.node.key.type === 'Identifier' && ['className', 'hoverClass', 'hoverClassName'].includes(path.node.key.name)) {
          startFlag = true
        }
      },
      exit (path) {
        if (path.node.key.type === 'Identifier' && ['className', 'hoverClass', 'hoverClassName'].includes(path.node.key.name)) {
          startFlag = false
        }
      }
    },
    StringLiteral: {
      enter (path, state) {
        if (startFlag) {
          path.node.value = replaceWxml(path.node.value, {
            keepEOL: true
          })
        }
      }
    },
    noScope: true
  })

  return generate(ast)
}
