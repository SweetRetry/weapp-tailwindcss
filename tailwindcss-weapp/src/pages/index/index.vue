<script setup lang="ts">
import { devDependencies } from '../../../package.json'
import BaseLayout from '@/components/BaseLayout.vue'
import FloatButton from '@/components/FloatButton.vue'
import { useSystemStore } from '@/stores'

const version = devDependencies.tailwindcss.slice(1)

const store = useSystemStore()

function openSearchBox() {

}

function toggleTheme() {
  if (store.theme === 'dark') {
    store.setTheme('light')
  }
  else {
    store.setTheme('dark')
  }
}

function go2ThemeDemo() {
  uni.navigateTo({
    url: '/pages/theme/index',
  })
}
</script>

<template>
  <BaseLayout>
    <up-navbar fixed bgColor="inherit" safeAreaInsetTop>
      <template #left>
        <view class="relative flex items-center">
          <view class="i-logos-tailwindcss-icon mr-2 h-[20.57px] w-[34px] shrink-0" />
          <view class="text-lg font-semibold">
            Tailwind CSS
          </view>
          <view
            class="relative bottom-0.5 ml-1 self-end rounded-full bg-sky-400/10 px-2 py-0.5 text-xs font-medium text-sky-600"
          >
            {{ version }}
          </view>
        </view>
      </template>
    </up-navbar>
    <FloatButton store-key="index-float-btn" :padding="[256, 32, 256, 32]">
      <view class="float-btn mb-3">
        <view class="i-mdi-flask-round-bottom-empty-outline text-sky-400 dark:text-sky-500" @click="go2ThemeDemo" />
      </view>
      <view class="float-btn mb-3" @click="toggleTheme">
        <view v-if="store.theme === 'dark'" class="i-mdi-white-balance-sunny text-sky-500" />
        <view v-else class="i-mdi-weather-night text-sky-400" />
      </view>
      <view class="float-btn" @click="openSearchBox">
        <view class="i-mdi-magnify text-sky-400 dark:text-sky-500" />
      </view>
    </FloatButton>
  </BaseLayout>
</template>
