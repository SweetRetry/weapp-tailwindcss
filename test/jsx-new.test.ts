import { jsxHandler as newJsxHandler } from '@/jsx/v2'

// import { jsxCasePath, createGetCase } from './util'

// const getCase = createGetCase(jsxCasePath)

// const putCase = createPutCase(jsxCasePath)

describe('jsxHandler', () => {
  // https://github.com/sonofmagic/weapp-tailwindcss-webpack-plugin/issues/53
  it('multiple file no end issue', async () => {
    const { code: code0 } = newJsxHandler(
      `_jsx(View, {
      className: 'border-[10px] border-[#098765] border-solid border-opacity-[0.44]'
    })`
    )
    expect(code0).toMatchSnapshot()
    const { code } = newJsxHandler(
      `import { replaceJs } from 'weapp-tailwindcss-webpack-plugin/replace'
    import './index.scss'
    import { jsx as _jsx } from 'react/jsx-runtime'
    import { jsxs as _jsxs } from 'react/jsx-runtime'
    import { Fragment as _Fragment } from 'react/jsx-runtime'`
    )
    expect(code).toMatchSnapshot()
  })
})
