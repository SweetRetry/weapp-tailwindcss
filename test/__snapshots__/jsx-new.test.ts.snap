// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsxHandler multiple file no end issue 1`] = `
"_jsx(View, {
  className: "border-_bl_10px_br_ border-_bl__h_098765_br_ border-solid border-opacity-_bl_0_d_44_br_"
});"
`;

exports[`jsxHandler multiple file no end issue 2`] = `
"import { replaceJs } from 'weapp-tailwindcss-webpack-plugin/replace';
import './index.scss';
import { jsx as _jsx } from 'react/jsx-runtime';
import { jsxs as _jsxs } from 'react/jsx-runtime';
import { Fragment as _Fragment } from 'react/jsx-runtime';"
`;
